modulemd/v2/modulemd-module.c

```diff
@@ -293,3 +309,39 @@ modulemd_module_get_stream_by_NSVC (ModulemdModule *self,
 
   return NULL;
 }
+
+
+void
+modulemd_module_add_translation (ModulemdModule *self,
+                                 ModulemdTranslation *translation)
+{
+  gsize i;
+  ModulemdModuleStream *stream = NULL;
+
+  if (!g_str_equal (modulemd_translation_get_module_name (translation),
+                    modulemd_module_get_module_name (self)))
+    return;
+
+  g_hash_table_replace (
+    self->translations,
+    g_strdup (modulemd_translation_get_module_stream (translation)),
+    g_object_ref (translation));
```

Comment:
You're taking a reference to the translation here, but making a copy of it for each individual stream that is added. That really doesn't make any sense to me. We should make a copy onto the highest-level location that will need to reference it, then take a reference on it at each lower level. So in this case, I think we need to take the copy *here* and then have `modulemd_module_stream_associate_translation()` take a reference on each stream that matches.

Also, when implementing the ModuleIndex we need to have the addition of a Translation or Defaults object cause the creation of the appropriate Module object, if a Stream has not already done so.

Reply:
The creation of the Module object I had already done: `get_or_create_module`

Reply:
We need to take a copy of the translation here (and use references to that copy within the associated streams). It's possible at the Index layer that we would get a Translation document passed in rather than built from YAML and we don't want to make the caller responsible for the memory (or for not changing it after it's referenced).